### ISSUE TITLE
eachMessage timeouts and health checks

### ISSUE THREAD

**Issue by pimpelsang** [tulios/kafkajs#447]
Does eachMessage have timeout?

The default session timeout is 30s,

What if my single processing takes longer? Right now I get consumer group rebalance as it seems heartbeat is only sent between messages? Perhaps heartbeat should be independent?

**Comment by Nevon** [comment]
`eachMessage` does not have a timeout. If you want that, you'll need to do it in userland. However, the heartbeat interval is essentially the "timeout", since if your `eachMessage` has not resolved within that time, the commit will be rejected.

The reason that heartbeat is not independent from the message processing is that it is intended as a healthcheck. If your consumer is not processing messages, it is not healthy, so the heartbeat should probably fail.

If you have some long running task that requires you to heartbeat intermittently, you can currently only do that via the `eachBatch` interface. However, #387 proposes a way to heartbeat outside of `eachBatch`, but it's not available yet.

**Comment by JaapRood** [comment]
Note that when you use `eachBatch`, which exposes a `heartbeat` method, you'll have to be sure that you're only sending heartbeats when processing is actually still healthy and didn't get stuck.  Setting a `setInterval` for it might keep the session alive, but generally is just disabling the failure detection mechanism completely, allowing your processing to get stuck.

**Comment by pimpelsang** [comment]
the heartbeat interval is very noisy as timeout as it causes all other partition consumers to rebalance as well. Right now with one bad message that processing takes over default 30s the entire consumer group falls into infinite rebalance loop.

**Comment by pimpelsang** [comment]
I can add timeout and Promise.race(), but it like feels something that the existing retry mechanism  should handle first and then maybe cause consumer group rebalance.

**Comment by JaapRood** [comment]
> the heartbeat interval is very noisy as timeout

I'm not sure exactly what you're saying, but I think there might be some confusion about `heartbeatInterval`. This **does not** cause automatic heartbeats, it only prevents heartbeats to happen more than whatever that is set to. You still have to manually call `heartbeat` (exposed in `eachBatch`), that value only prevents from sending hundreds of heartbeats per second. If `heartbeat` is called in a timely manner it should _prevent_ a rebalance.

An example situation where we needed this was when processing a message we had to "wait" for something else to happen first, so we polled it's status every 3 seconds. After polling 10 times, the 30 second session time-out was reached and a rebalance occurred. To fix it, we called `hearbeat` after every time we polled the other resource for an update.
